staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/marshal_test.go

```diff
@@ -169,3 +432,105 @@ func TestJSONUnderlyingArrayReuse(t *testing.T) {
 		t.Errorf("unexpected mutation, got %s want %s", got, want)
 	}
 }
+
+func unmarshalJSONTest[T Testable[T]](input []byte, wantErr bool, expected T) func(t *testing.T) {
+	return func(t *testing.T) {
+		var actualDecoded T
+		actualDecoded = reflect.New(reflect.TypeOf(actualDecoded).Elem()).Interface().(T)
+		err := actualDecoded.UnmarshalJSON(input)
+		if (err != nil) != wantErr {
+			if wantErr {
+				t.Fatal("expected error")
+			}
+			t.Fatalf("unexpected error: %v", err)
+		}
+		if err == nil && !cmp.Equal(expected, actualDecoded) {
+			t.Error("unexpected decoded value")
+			t.Fatalf(cmp.Diff(expected, actualDecoded))
+		}
+	}
+}
+
+func roundTripJSONTest[T Testable[T]](input T, expectedEncoded []byte, expectedDecoded T) func(t *testing.T) {
+	return func(t *testing.T) {
+		input := input.DeepCopy()
+		actualEncoded, err := input.MarshalJSON()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if bytes.Compare(expectedEncoded, actualEncoded) != 0 {
+			t.Error("unexpected encoded value")
+			t.Fatal(cmp.Diff(string(expectedEncoded), string(actualEncoded)))
+		}
+		var actualDecoded T
+		actualDecoded = reflect.New(reflect.TypeOf(actualDecoded).Elem()).Interface().(T)
+		err = actualDecoded.UnmarshalJSON(actualEncoded)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if reflect.ValueOf(expectedDecoded).IsZero() {
+			expectedDecoded = input
+		}
+		if !cmp.Equal(expectedDecoded, actualDecoded) {
+			t.Error("unexpected decoded value")
+			t.Fatalf(cmp.Diff(expectedDecoded, actualDecoded))
+		}
+	}
+}
+
+func roundTripCBORTest[T Testable[T]](input T, expectedEncoded []byte, expectedDecoded T) func(t *testing.T) {
+	return func(t *testing.T) {
+		input := input.DeepCopy()
+		actualEncoded, err := input.MarshalCBOR()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if bytes.Compare(expectedEncoded, actualEncoded) != 0 {
+			t.Error("unexpected encoded value")
+			t.Fatal(cmp.Diff(expectedEncoded, actualEncoded))
+		}
+		var actualDecoded T
+		actualDecoded = reflect.New(reflect.TypeOf(actualDecoded).Elem()).Interface().(T)
+		err = actualDecoded.UnmarshalCBOR(actualEncoded)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if reflect.ValueOf(expectedDecoded).IsZero() {
+			expectedDecoded = input
+		}
```

Comment:
Instead of having a special case for this in the test logic, I think it'd be clearer to explicitly pass in the same object twice during test construction.